### DISCUSSION TITLE
sys.excepthook and pudb

### DISCUSSION THREAD

**Discussion by presto-dev** [inducer/pudb#549] [Troubleshooting]
Below is an example test script that I've used a lot. I started a new python venv with python3.10 and suddenly it no longer works.

I tested this with a new 
`python3.10 -m venv py310_test`
`source py310_test/bin/activate`
`pip install requirements` # see list below

under my old venv (python3.7) pudb would auto launch when oh_no() is executed, allowing me to inspect the stack

under the new venv it simply dumps the stack and terminates the script and i lose all the variables

please advise how to troubleshoot. i still have my old venv if there's anything in there worth looking at
```python
#!/usr/bin/env python


import traceback
import pudb as debugger
import sys


def main():
    sys.excepthook = drop_debugger
    import pudb
    pudb.pm()
    oh_no()
    return


def drop_debugger(type, value, tb):
    traceback.print_exception(type, value, tb)
    debugger.pm()


def oh_no():
    raise Exception("oh no!")


if __name__ == "__main__":
    main()
```

requirements below generated by pip-chill

`aenum
autopep8
black
ciscoconfparse
dos2unix
flake8
humanize
ipcalc
ipython
junos-eznc
neovim
netmiko
npyscreen
pandas
password-strength
pip-chill
protobuf
ptpython
pudb
pycrypto
pymysql
requests
rich
setproctitle
si-prefix
sshtunnel
tabulate
termcolor
terminal
tqdm
upass
urwidtrees
xml2dict
xmltodict
yarp
`

**Comment by inducer** [comment]
You could try to match the pudb version from the old venv in the new one, to see if it's a change in pudb that makes the difference.

**Comment by presto-dev** [comment]
@inducer excellent tip

Following your advice, I spun up a new ubuntu22 virt-server just to rule out anything weird on my system

created 2 venv environments

```
user@ubuntu22-server:~$ python3 -m venv old_pudb
user@ubuntu22-server:~$ python3 -m venv new_pudb
```

next 

```
user@ubuntu22-server:~$ source new_pudb/bin/activate
(new_pudb) user@ubuntu22-server:~$ pip install pudb
```

observed pudb-2022.1.2 is installed, along with dependancies

```
Collecting pudb
  Downloading pudb-2022.1.2.tar.gz (219 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 219.7/219.7 KB 1.8 MB/s eta 0:00:00
```

./pudb_test.py (script from my original post) dumps stack trace, no pudb interface is presented

`(new_pudb) user@ubuntu22-server:~$ pip freeze > requirements.txt`

edited requirements.txt
 
$ cat requirements.txt
pudb==2019.1

change venv

```
(new_pudb) user@ubuntu22-server:~$ deactivate
user@ubuntu22-server:~$ source old_pudb/bin/activate
(old_pudb) user@ubuntu22-server:~$ pip install -r requirements.txt
Collecting pudb==2019.1
  Downloading pudb-2019.1.tar.gz (57 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.7/57.7 KB 824.7 kB/s eta 0:00:00

```

when I run ./pudb_test.py, the exception raised by oh_no() causes script to pause and pudb interface is presented


I confirmed this is a reproducible experience on a fresh system standardized system. What I don't know if this is expected behavior and I simply don't understand, or if this is indeed a bug.

I hope this is helpful. I can continue using  pudb-2019.1 in python3.10 so this is not a show-stopper for me

Thanks very much for the suggestion.

**Comment by mvanderkamp** [comment]
Quick note on github markdown- you need triple backticks to define a multiline code block. See https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Reply by presto-dev** [reply]
thank you, i cleaned it up

**Comment by mvanderkamp** [comment]
Thanks for cleaning that up! I was able to replicate the behaviour you're seeing, so I bisected. The commit which changed this behaviour was f77332c5586dccb96cd85b9b92477ea6f749dd65

**Reply by presto-dev** [reply]
excellent, thank you for the support
